shared/subscribers/service_test.go

```diff
@@ -0,0 +1,148 @@
+package subscribers
+
+import (
+	"context"
+	"testing"
+
+	"code.vegaprotocol.io/vega/core/events"
+	dtypes "code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/logging"
+	eventspb "code.vegaprotocol.io/vega/protos/vega/events/v1"
+	"code.vegaprotocol.io/vega/shared/subscribers/mocks"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSlowConsumerIsDisconnected(t *testing.T) {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	ctrl := gomock.NewController(t)
+
+	broker := mocks.NewMockBroker(ctrl)
+	broker.EXPECT().Subscribe(gomock.Any())
+	broker.EXPECT().Unsubscribe(gomock.Any())
+
+	maxBufferSize := 3
+	testStreamSubscription := TestStreamSubscription{events: make(chan []*eventspb.BusEvent)}
+
+	var serviceContext context.Context
+	s := NewService(logging.NewTestLogger(), broker, maxBufferSize)
+	out, _ := s.ObserveEventsOnStream(ctx, 2, func(passedCtx context.Context) StreamSubscription {
+		serviceContext = passedCtx
+		return testStreamSubscription
+	})
+
+	testStreamSubscription.events <- []*eventspb.BusEvent{
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-1",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-2",
+		}).StreamMessage(),
+	}
+
+	events1 := <-out
+	assert.Equal(t,
+		[]*eventspb.BusEvent{
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-1"}).StreamMessage(),
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-2"}).StreamMessage(),
+		},
+		events1)
+
+	testStreamSubscription.events <- []*eventspb.BusEvent{
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-3",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-4",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-5",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-6",
+		}).StreamMessage(),
+	}
+
+loop:
+	for {
+		select {
+		case <-serviceContext.Done():
```

Comment:
This looks like a really elaborate way to write the equivalent of:

```
<-ctx.Done()
// or
for _ := range <-ctx.Done() {
    break
}
```

don't really see a point to the `loop` label either

Reply:
ah yeah, sorted